core/domain/story_services.py

```diff
@@ -559,6 +559,28 @@ def get_completed_node_ids(user_id, story_id):
     return progress_model.completed_node_ids if progress_model else []
 
 
+def get_latest_completed_node_ids(user_id, story_id):
+    """Returns the  ids of the nodes lastest completed in the story.
+
+    Args:
+        user_id: str. ID of the given user.
+        story_id: str. ID of the story.
+
+    Returns:
+        list(str). List of the node ids lastest completed in story. If length
+        is larger than 3, return the lastest three of them. If length is smaller
+        or equal to 3, return all of them.
+    """
+    progress_model = user_models.StoryProgressModel.get(
+        user_id, story_id, strict=False)
+
+    if not progress_model:
+        return []
+
+    num_of_nodes = min(len(progress_model.completed_node_ids), 3)
+    return progress_model.completed_node_ids[-num_of_nodes:]
```

Comment:
Do we have any controls to ensure that progress_model.completed_node_ids is sorted according to how you want to retrieve this data? If not, we should add those checks. (Don't just assume that because it's implemented a certain way now, it will be implemented like that in the future -- things may change and devs need to know what constraints the list operates under, so add comments and automated tests to guard as needed.)

Also is it possible for a user to re-complete an old node? Need to handle that too.

Reply:
I am pretty sure the completed node ids are sorted in the order that when it was completed. You can see that it is stored as a list in user/gae_models.py, and in record_completed_node_in_story_context (story_services.py), the newly completed node will be appended at the end of the list.

Here is the code for reference:
`progress_model = user_models.StoryProgressModel.get_or_create(
        user_id, story_id)`
    `if node_id not in progress_model.completed_node_ids:
        progress_model.completed_node_ids.append(node_id)
        progress_model.put()`

Reply:
1) They currently happen to be, but I'm not sure that that's a guarantee the model makes to its users. If we are requiring this then it should be part of the progress model field description and it should be enforced somehow. E.g. if a developer decides to change the behaviour of "re-complete an older exploration node" in the future to put that node at the end of the list (and remove the existing entry), that should cause tests to fail.

2) That said, there's a conceptual difference (especially if stories end up branching) between "last three explorations completed" and "most advanced / latest three explorations completed", although these are currently the same because our stories are linear. Which concept are you going with? It seems to me that we should be looking for the last exps in order of the story path, not the last exps that the student played.

Reply:
I thought the design of story viewer would be that it is restricted to play nodes in order, so the latest completed nodes would always be the last three nodes. I think this would make more sense because lessons are supposed to be learned in order. I think it would be helpful if I can have more information about the design of story viewer.

This is my thinking now:
exp1 -- exp2 -- exp3 -- review1 -- exp4 -- exp5 -- review2 , etc. They will be displayed as a list or table in the story viewer, and the later ones will be disabled before completing the ones before them.

Reply:
Yes, that is what I expect as well. But UI decisions change in the future whereas storage models remain constant, so we should be clear about what we're assuming about the storage model.

So, what I'm saying is, can you retrieve the progress and find the 3 collection IDs that are "furthest along" in the collection path? Assume that the list is unsorted. That would be better than assuming that they are the last three in the list, and would help future-proof your code as requirements of other parts of the system change.

Reply:
Done.